marshal.go

```diff
@@ -1454,25 +1460,40 @@ func unmarshalDuration(info TypeInfo, data []byte, value interface{}) error {
 	return unmarshalErrorf("can not unmarshal %s into %T", info, value)
 }
 
-func decVints(data []byte) (int32, int32, int64) {
-	month, i := decVint(data)
-	days, j := decVint(data[i:])
-	nanos, _ := decVint(data[i+j:])
-	return int32(month), int32(days), nanos
+func decVints(data []byte) (int32, int32, int64, error) {
+	month, i, err := decVint(data, 0)
+	if err != nil {
+		return 0, 0, 0, errors.Wrapf(err, "failed to extract month")
+	}
+	days, i, err := decVint(data, i)
+	if err != nil {
+		return 0, 0, 0, errors.Wrapf(err, "failed to extract days")
+	}
+	nanos, _, err := decVint(data, i)
+	if err != nil {
+		return 0, 0, 0, errors.Wrapf(err, "failed to extract nanoseconds")
+	}
+	return int32(month), int32(days), nanos, err
 }
 
-func decVint(data []byte) (int64, int) {
-	firstByte := data[0]
+func decVint(data []byte, start int) (int64, int, error) {
```

Comment:
@martin-sucha , agree it looks ugly as hell, but there is point in doing that.
Since it is serialization code, it has impact on performance, returning buffer implies that `slice` (not an array) is going to be recreated, which is not realy necessary and could be avoided.

Reply:
@martin-sucha , agree it looks ugly, but there is point in doing that.
Since it is serialization code, it has impact on performance, returning buffer implies that `slice` (not an array) is going to be recreated, which is not realy necessary and could be avoided.
 

> Please update AUTHORS and CHANGELOG as per [CONTRIBUTING.md](https://github.com/gocql/gocql/blob/master/CONTRIBUTING.md#minimum-requirement-checklist)

Done

Reply:
```
func decVint(data []byte) (int64, []byte, error) {
	if len(data) == 0 {
		return 0, nil, errors.New("unexpected eof")
	}
	firstByte := data[0]
	if firstByte&0x80 == 0 {
		return decIntZigZag(uint64(firstByte)), data[1:], nil
	}
	numBytes := bits.LeadingZeros32(uint32(^firstByte)) - 24
	ret := uint64(firstByte & (0xff >> uint(numBytes)))
	if len(data) < numBytes+1 {
		return 0, nil, fmt.Errorf("data expect to have %d bytes, but it has only %d", numBytes+1, len(data))
	}
	for i := 0; i < numBytes; i++ {
		ret <<= 8
		ret |= uint64(data[i+1] & 0xff)
	}
	return decIntZigZag(ret), data[numBytes+1:], nil
}
```

Gave results:
```
goos: linux
goarch: amd64
pkg: github.com/gocql/gocql
cpu: 12th Gen Intel(R) Core(TM) i9-12900HK
BenchmarkDecVints-20             1000000              1034 ns/op               0 B/op          0 allocs/op
```

While current solution showed following results:
```
goos: linux
goarch: amd64
pkg: github.com/gocql/gocql
cpu: 12th Gen Intel(R) Core(TM) i9-12900HK
BenchmarkDecVints-20             1221241               970.1 ns/op             0 B/op          0 allocs/op
PASS
ok      github.com/gocql/gocql  3.344s
```

Test code:
```
func BenchmarkDecVints(b *testing.B) {
	data := []byte("\x89\xa2\xc3\xc2\x9a\xe0F\x91\x06")
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		for x := 0; x < 100; x++ {
			_, _, _, _ = decVints(data)
		}
	}
}
```

Reply:
Thank you, I stand corrected. I can reproduce the result:

```
goos: linux
goarch: amd64
pkg: github.com/gocql/gocql
cpu: Intel(R) Core(TM) i7-10850H CPU @ 2.70GHz
            │     old     │                new                 │
            │   sec/op    │   sec/op     vs base               │
DecVints-12   1.862µ ± 5%   1.975µ ± 6%  +6.04% (p=0.001 n=10)
```

Also with benchmark with exported variables:

```go
var (
	B1, B2 int32
	B3     int64
	B4     error
)

func BenchmarkDecVints(b *testing.B) {
	data := []byte("\x89\xa2\xc3\xc2\x9a\xe0F\x91\x06")
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		for x := 0; x < 100; x++ {
			B1, B2, B3, B4 = decVints(data)
		}
	}
}
```

```
goos: linux
goarch: amd64
pkg: github.com/gocql/gocql
cpu: Intel(R) Core(TM) i7-10850H CPU @ 2.70GHz
            │     old     │                new                 │
            │   sec/op    │   sec/op     vs base               │
DecVints-12   1.914µ ± 1%   2.042µ ± 1%  +6.66% (p=0.000 n=10)

```

Reply:
@martin-sucha , did you compare it to `master` ? If so, you should not do that, since `master` code luck of validations, and by definition going to be faster, but less safer, you have to add validations to the master and then compare it to the code from this PR

Reply:
Sorry for confustion. In the output above, old is `func decVints(data []byte) (int32, int32, int64, error)`, new is `func decVint(data []byte) (int64, []byte, error)`. What I was saying is that I run the same benchmark as you and got the same result.